src/Shared/LoadedType.cs

```diff
@@ -36,55 +38,78 @@ internal LoadedType(Type type, AssemblyLoadInfo assemblyLoadInfo, Assembly loade
             ErrorUtilities.VerifyThrow(type != null, "We must have the type.");
             ErrorUtilities.VerifyThrow(assemblyLoadInfo != null, "We must have the assembly the type was loaded from.");
 
-            _type = type;
+            try
+            {
+                Type t = Type.GetType(type.AssemblyQualifiedName);
+                if (t.Assembly.Location.Equals(loadedAssembly.Location, StringComparison.OrdinalIgnoreCase))
+                {
+                    _type = t;
+                }
+            }
+            catch (Exception) { }
+            _type ??= type;
             _assembly = assemblyLoadInfo;
             _loadedAssembly = loadedAssembly;
 
-            CheckForHardcodedSTARequirement();
-            HasLoadInSeparateAppDomainAttribute();
-            HasSTAThreadAttribute();
+            HasSTAThreadAttribute = CheckForHardcodedSTARequirement();
+            if (loadedAssembly is null)
+            {
+                HasLoadInSeparateAppDomainAttribute = this.Type.GetTypeInfo().IsDefined(typeof(LoadInSeparateAppDomainAttribute), true /* inherited */);
+                HasSTAThreadAttribute = this.Type.GetTypeInfo().IsDefined(typeof(RunInSTAAttribute), true /* inherited */);
+                IsMarshalByRef = this.Type.GetTypeInfo().IsMarshalByRef;
+            }
+            else
+            {
+#if !NET35
+                Type t = type;
+                while (t is not null)
+                {
+                    if (CustomAttributeData.GetCustomAttributes(t).Any(attr => attr.AttributeType.Name.Equals("LoadInSeparateAppDomainAttribute")))
+                    {
+                        HasLoadInSeparateAppDomainAttribute = true;
+                    }
+
+                    if (CustomAttributeData.GetCustomAttributes(t).Any(attr => attr.AttributeType.Name.Equals("RunInSTAAttribute")))
```

Comment:
Tiny perf optimization possibility: walk the list only once and check for both attributes.